src/TimeoutMigrationTool/Asp/AspTimeoutsSource.cs

```diff
@@ -0,0 +1,668 @@
+        {
+            var tableClient = CreateCloudTableClient();
+
+            var migrationTable = GetMigrationTable(tableClient);
+
+            var query = new TableQuery<PartialMigratedTimeoutDataEntityWithBatchState>()
+                .Where(GenerateFilterCondition("PartitionKey", QueryComparisons.Equal,
+                    batchNumber.ToString(CultureInfo.InvariantCulture)));
+
+            TableContinuationToken token = null;
+            CancellationToken queryCancellationToken = CancellationToken.None;
+
+            do
+            {
+                // will issue a table scan for now
+                var seg = await migrationTable.ExecuteQuerySegmentedAsync(
+                        query: query,
+                        token: token,
+                        requestOptions: null,
+                        operationContext: null,
+                        cancellationToken: CancellationToken.None)
+                    .ConfigureAwait(false);
+                token = seg.ContinuationToken;
+
+                // with batching we can at least efficiently update
+                var batch = new TableBatchOperation();
+                foreach (var t in seg.Results)
+                {
+                    t.BatchState = batchState;
+
+                    // we don't want to preserve the etag and we always want to win
+                    t.ETag = "*";
+
+                    // the batch can have max 100 items and max 4 MB of data
+                    // we don't need to check the size since we are only getting partition and row key as well as batch state
+                    // the partition key for all operations in the batch has to be the same
+                    if (batch.Count == 100)
+                    {
+                        await migrationTable.ExecuteBatchAsync(batch);
+                        batch.Clear();
+                    }
+
+                    batch.Add(TableOperation.Merge(t));
+                }
+
+                if (batch.Count > 0)
+                {
+                    await migrationTable.ExecuteBatchAsync(batch);
+                }
+
+            } while (token != null && !queryCancellationToken.IsCancellationRequested);
+
+            var toolState = await TryLoadToolStateEntity(tableClient);
+            toolState.CurrentBatchNumber = nextBatch;
+
+            var toolStateTable = await GetAndCreateToolStateTableIfNotExists(tableClient);
+            await toolStateTable.ExecuteAsync(TableOperation.Merge(toolState));
+        }
+
+        public async Task<IReadOnlyList<EndpointInfo>> ListEndpoints(DateTime cutOffTime)
+        {
+            var tableClient = CreateCloudTableClient();
+
+            var endpointTimeoutTable = await GetEndpointTimeoutTable(endpointNameToBelisted, tableClient);
+
+            var cuttOffTimeAsPartitionKeyScope = cutOffTime.ToString(partitionKeyScope);
+            var upperLimitForCuttOffToFilterGuidEntriesOut = cutOffTime.AddYears(100).ToString(partitionKeyScope);
```

Comment:
What does this do? Why are there entries further in the future than the cutoff that we care about?

Reply:
Added a comment. will push soon